feat(query): Observable interop

backbone87 opened the pull request:
Fixes #176

## Proposed Changes

- Added `QueryApi.lines(...)` and `QueryApi.rows(...)` which return cold observables that are interoperable with the TC39 observable proposal and allow consumption of those observables by other libs like rxjs & zen-observables

## Checklist

  - [x] CHANGELOG.md updated
  - [x] A test has been added if appropriate
  - [x] `yarn test` completes successfully
  - [x] Commit messages are in [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages)
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

backbone87 commented:
example usage:
```ts
import {InfluxDB} from '@influxdata/influxdb-client'
import {from} from 'rxjs'
import {map} from 'rxjs/operators'

const influx = new InfluxDB({
  url: 'http://localhost:9999',
  token: '...',
})
const queryApi = influx.getQueryApi('test').with({ defaultVariables: { bucket: 'test' } })

const query = `
  from(bucket: v.bucket)
    |> range(start: -1d)
    |> filter(fn: (r) => contains(value: r.object_id, set: v.objectIds)
    )
    |> last()
    |> keep(columns: ["object_id", "_time", "_field", "_value"])
`
const variables = {objectIds: ['MY_OBJECT_ID']}

from(queryApi.rows(query, variables))
  .pipe(map(({values, tableMeta}) => tableMeta.toObject(values)))
  .forEach(console.log)
```

sranka commented on packages/core/src/impl/ObservableQuery.ts:
```diff
@@ -0,0 +1,79 @@
+  ) {}
+
+  public subscribe(
+    observerOrNext?: Partial<Observer<T>> | ((value: T) => void)
```
what about error and complete optional parameters?

backbone87 replied:
resolving here, see the discussion on the observable typings

sranka commented on packages/core/src/impl/QueryApiImpl.ts:
```diff
@@ -23,27 +27,69 @@ export class QueryApiImpl implements QueryApi {
+  private execute(
+    query: string,
+    extern: unknown | undefined,
+    {org, type, gzip}: QueryOptions,
```
What is the benefit of having QueryOptions parameter herein? We could destructure this.options in the fn body.

backbone87 replied:
the options of an observable query should not change after its creation, thats why the `createExecutor` copies the current options and then passes this copy to `execute`

sranka commented on packages/core/src/observable/types.ts:
```diff
@@ -0,0 +1,26 @@
+export interface Observable<T> {
+  subscribe(): Subscription
+  subscribe(observer: Partial<Observer<T>>): Subscription
+  subscribe(next: (value: T) => void): Subscription
```
`onError? : Function, onComplete? : Function` are missing comparing to [TC39 observable proposal](https://github.com/tc39/proposal-observable)

backbone87 replied:
yeah i looked here https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/zen-observable/index.d.ts#L36 insteadof here https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/zen-observable/index.d.ts#L46

and i also had type errors with rxjs, because they introduce `null` on some props.

i will add this back in

sranka commented on packages/core/src/observable/types.ts:
```diff
@@ -0,0 +1,26 @@
+/* Observable interop typing. Taken from https://github.com/ReactiveX/rxjs */
```
There is no such type in https://github.com/ReactiveX/rxjs/blob/master/src/internal/types.ts

backbone87 replied:
Subscribable -> Observable: the rxjs Subscribable is the actual Observable from the TC39 proposal, the rxjs Observable interface has other stuff included
Unsubscribable + { closed: boolean } -> Subscription: 
PartialObserver -> Partial<Observer>: the PartialObserver of rxjs require at least 1 observer method, but this is not required for us and the typing simplifies on our side and its still interoperable

sranka replied:
Understood. Please remove this comment, since there is nothing taken from rxjs.

backbone87 commented:
> Wouldn't it be better to create 2 PRs? One with the Observable interop and another with (optional) Variables in all query methods.

we can do this, but lets draft the implementation together to a stable point here, and i create separate MRs later

sranka commented on packages/core/src/util/flux.ts:
```diff
@@ -0,0 +1,179 @@
+/*
```
Is there any doc that better explains how this magic works? Where did you get it?

backbone87 replied:
check this link in the comment ;) https://github.com/influxdata/influxdb/issues/16109#issuecomment-561697800
this + some reverse engineering of the flux AST via https://v2.docs.influxdata.com/v2.0/api/#operation/PostQueryAst

backbone87 replied:
"extern" is basically like a automatically imported file into the query, but we have to pass the flux AST insteadof source text

sranka replied:
Good job on your side ... still, there might be problems with backward compatibility, once we introduce something that does not seem to be stable enough. There is no documentation and the things that are reverse-engineered might change easily.  I am also not sure whether this is working against InfluxDB 1.8, which must be supported as well. We should also plan support for query parameters in all client libraries once it is stable, so they are in sync. I am now not sure whether this is a good step forward, maybe that we will need other stakeholders herein.

2 PRs would play better herein. The observable part of this PR is fine with no side-effects and ready to be merged soon. The variable part would require some argumentation to proceed.

sranka commented on packages/core/src/QueryApi.ts:
```diff
@@ -40,6 +53,22 @@ export default interface QueryApi {
+   * @param query the query text in the format specifed in `QueryOptions.type`
+   * @param variables variables to add to the query, when `QueryOptions.type` is "flux"
+   */
+  rows(query: string, variables?: Variables): Observable<Row>
```
naming: My preference herein would be QueryParameters, since such a naming convention (parameter) is used in the majority of existing libraries that do queries. A variable (to me) means something that can change ... and to be semantically correct ... these are variable bindings.

backbone87 replied:
will change, had it called "Params" before :)
`Parameters` (shadows builtin TS type)
`Bindings`
`ParameterBindings`
`VariableBindings`

whats your preference?

backbone87 replied:
actually these are not bindings, because the passed values get serialized into a flux AST immediately. so changing a passed in value after the call, wouldnt affect the query.
guess i will go with `parameters: QueryParameters`

backbone87 commented on packages/core/src/QueryApi.ts:
```diff
@@ -40,6 +53,22 @@ export default interface QueryApi {
+   * @param query the query text in the format specifed in `QueryOptions.type`
+   * @param variables variables to add to the query, when `QueryOptions.type` is "flux"
+   */
+  lines(query: string, variables?: Variables): Observable<string>
```
i didnt do this yet, because i thought the order of parameters felt strange then:
`queryLines(query, consumer, variables)`

sranka replied:
Yes, the callback should be the last by convention. It can be realized in a backward-compatible way.

*interface*:
```js
  queryLines(query: string, consumer: CommunicationObserver<string>): void
  queryLines(
    query: string,
    parameters: QueryParameters,
    consumer: CommunicationObserver<string>
  ): void
```
*implementation*:
```
  queryLines(
    query: string,
    params: any | CommunicationObserver<string>,
    consumer?: CommunicationObserver<string>
  ): void {
    if (arguments.length === 2) {
      consumer = params as CommunicationObserver<string>
    }
...
// consumer as CommunicationObserver<string> 
```

backbone87 commented:
i have removed the query parameters from this PR and split it into https://github.com/influxdata/influxdb-client-js/pull/178 (also copied some conversation over)

whats missing for this PR are integration tests with at least one other observable library

sranka commented on packages/core/src/observable/types.ts:
```diff
@@ -0,0 +1,35 @@
+  subscribe(next: ObserverNext<T>): Subscription
+  subscribe(
+    next: ObserverNext<T>,
+    error: ObserverError,
```
error is optional, so it should be  `error?: ObserverError` in order to align with tc39; I can also imagine that all params are optional, so that the interface is just  
```js
subscribe(observer: Partial<Observer<T>>): Subscription
subscribe(
    next?: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription
```

backbone87 replied:
if you dont pass error the previous signature matches:
`subscribe(next: ObserverNext<T>): Subscription`

there should be "no holes" (undefineds) in the arg list, which your proposed signature would allow:
```ts
subscribe(
    next?: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription

subscribe(undefined, undefined, () => console.log('yay')) // this is not required by the spec and deprecated in rxjs
```

backbone87 replied:
https://tc39.es/proposal-observable/#observable-prototype-subscribe

its actually a little bit confusing:
the described algorithmn allows everything in the first position:
- object -> observer, ignore rest of the args
- callable -> its the next function, check other "optional" args (so `subscribe(console.log, undefined, () => console.log('yay'))` is allowed)
- everything else -> empty observer (this includes `subscribe(undefined, undefined, () => console.log('yay'))`)

so insteadof:
```ts
export interface Observable<T> {
  subscribe(): Subscription
  subscribe(observer: Partial<Observer<T>>): Subscription
  subscribe(next: ObserverNext<T>): Subscription
  subscribe(
    next: ObserverNext<T>,
    error: ObserverError,
    complete?: ObserverComplete
  ): Subscription
}
```
its actually
```ts
export interface Observable<T> {
  subscribe(): Subscription
  subscribe(observer: Partial<Observer<T>>): Subscription
  subscribe(
    next: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription
}
```

sranka replied:
so it is then identical to what I suggested
```
subscribe(observer: Partial<Observer<T>>): Subscription
subscribe(
    next?: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription
```

backbone87 replied:
i did go with
```ts
  subscribe(observer: Partial<Observer<T>>): Subscription
  subscribe(
    next: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription
```

- the first parameter is required
- the first parameter can be anything, but in TS we can be more smart and only allow stuff that makes sense

```ts
subscribe(
    next?: ObserverNext<T>,
    error?: ObserverError,
    complete?: ObserverComplete
  ): Subscription
```
while its a valid signature according to the spec, it would produce an empty observer, when calling `subscribe(undefined, console.log)`, which is unexpected and therefore disallowed

backbone87 commented:
added rxjs intergration tests (`cd ./packages/core && yarn test:integration` needs to be added to the CI/global scripts)
added example

you can take over from here i think

backbone87 commented:
TODO: add `export { Row } from './QueryApi'` in index.ts
(pls you do this, i dont want to interfer anymore while you finish up this PR)

sranka commented:
@backbone87 I added my touches herein (FYI):
- finished the rxjs example 
   - rxjs renamed to rxjs-query, since there could be also an rxjs write
   - added rxjs to examples package.json, since examples should be also built standalone
   - make the example executable
- updated CHANGELOG with the PR
- added more tests to cover the code
- repaired test failures in corner cases
   - unsubscribe closes the subscription
- re-exported Row from the main index.ts

Please check that you have signed CLA in the PR CheckList above, so I could do a final review, approval and merge. Thank you for your contribution!

backbone87 commented:
> Please check that you have signed CLA

working on that, checking back with my employer (shouldnt be a problem)

backbone87 commented:
CLA signed

sranka commented:
It is available in nightly builds, verified with rxjs-query.ts with modified package.json

```js
    "@influxdata/influxdb-client-apis": "1.3.0-nightly.3+f46c96b",
    "@influxdata/influxdb-client": "1.3.0-nightly.3+f46c96b"
```
